compiler/src/dotty/tools/dotc/typer/Migrations.scala

```diff
@@ -0,0 +1,108 @@
+package dotty.tools
+package dotc
+package typer
+
+import core.*
+import ast.*
+import Contexts.*
+import Types.*
+import Flags.*
+import Names.*
+import StdNames.*
+import Symbols.*
+import Trees.*
+import ProtoTypes.*
+import Decorators.*
+import config.MigrationVersion
+import config.Feature.{sourceVersion, migrateTo3}
+import config.SourceVersion.*
+import reporting.*
+import NameKinds.ContextBoundParamName
+import rewrites.Rewrites.patch
+import util.Spans.Span
+import rewrites.Rewrites
+
+/** A utility trait containing source-dependent deprecation messages
+ *  and migrations.
+ */
+trait Migrations:
+  this: Typer =>
```

Comment:
I am not convinced we should make this one a mixin. The one use case right now for `typedAsFunction`, all other functions that are here and could go in here seem to be static.

I would suggest making `Migrations` an `object` and redefining `Typer.typedAsFunction` as
```scala
def typedAsFunction(tree: untpd.PostfixOp, pt: Type)(using Context): Tree =
    migrate(Migrations.typedAsFunction(tree, pt), throw new AssertionError("can't retype a PostfixOp"))
```